File: frontend/panel/src/auth/permissions.ts
/**
 * Client-side permission model (contract C2/C7). Phase 2 mirrored a hardcoded
 * role→permission matrix; Phase 3 makes roles DB-backed and editable, so the
 * authoritative effective set is the `perms` claim the backend embeds in every
 * access token (see `internal/auth/tokens.go`). We decode that claim and gate
 * the UI against *permission strings* — never role names (C2). Admin carries the
 * wildcard `*` (allow-all).
 *
 * The panel is a convenience layer only: the server re-checks every permission,
 * so a hidden button is not a security boundary. If a token cannot be decoded
 * (e.g. a stale Phase-2 session) we fall back to the frozen builtin matrix keyed
 * on the manager's role string so the UI never silently opens up or locks out.
 */

// Module verb permissions used to gate whole screens/actions.
export const PERM_SUBSCRIBERS_VIEW = 'subscribers.view'
export const PERM_PROFILES_VIEW = 'profiles.view'
export const PERM_NAS_VIEW = 'nas.view'
export const PERM_POOLS_VIEW = 'pools.view'
export const PERM_LIVE_VIEW = 'live.view'
export const PERM_REPORTS_VIEW = 'reports.view'
export const PERM_AUDIT_VIEW = 'audit.view'
export const PERM_MANAGERS_VIEW = 'managers.view'
export const PERM_MANAGERS_CREATE = 'managers.create'
export const PERM_MANAGERS_EDIT = 'managers.edit'
export const PERM_MANAGERS_DELETE = 'managers.delete'
export const PERM_VOUCHERS_VIEW = 'vouchers.view'
export const PERM_VOUCHERS_CREATE = 'vouchers.create'
export const PERM_MONITORING_VIEW = 'monitoring.view'
export const PERM_MONITORING_EDIT = 'monitoring.edit'
export const PERM_SETTINGS_VIEW = 'settings.view'
export const PERM_SETTINGS_EDIT = 'settings.edit'
export const PERM_LICENSE_MANAGE = 'license.manage'
export const PERM_BACKUPS_VIEW = 'backups.view'
/** v2-2 (FR-79.2/77.1): generalizes card_payments.verify to every method. */
export const PERM_PAYMENT_TICKETS_VERIFY = 'payment_tickets.verify'
export const PERM_PAYMENT_PROVIDERS_MANAGE = 'payment_providers.manage'
export const PERM_SUBSCRIBERS_CREATE = 'subscribers.create'
export const PERM_NAS_EDIT = 'nas.edit'
/** v2 phase 7 (FR-87.1): one-click panel update, admin-only by default. */
export const PERM_SYSTEM_UPDATE = 'system.update'

// Bare action permissions granted independently of module verbs (C2/C7).
export const PERM_DISCONNECT = 'disconnect'
export const PERM_EXPORT = 'export'
export const PERM_RENEW = 'renew'
export const PERM_TOPUP = 'topup'
export const PERM_REFUND = 'refund'
/** v2 phase 4 (FR-68.3): rate creation is admin-only, an append-only audit trail. */
export const PERM_CURRENCY_RATES_MANAGE = 'currency_rates.manage'
/** v2 phase 9 (FR-73/74): overheads and reseller wholesale pricing are admin-only. */
export const PERM_OVERHEADS_MANAGE = 'overheads.manage'
export const PERM_RESELLER_PRICES_MANAGE = 'reseller_prices.manage'

const WILDCARD = '*'
const ROLE_ADMIN = 'admin'

// Fallback matrix — byte-for-byte the backend's builtin rolePermissions map,
// used only when a token has no decodable `perms` claim. Admin is allow-all and
// intentionally absent (see hasPermission).
const ROLE_PERMISSIONS: Record<string, ReadonlySet<string>> = {
  operator: new Set([
    PERM_SUBSCRIBERS_VIEW,
    'subscribers.create',
    'subscribers.edit',
    PERM_PROFILES_VIEW,
    PERM_NAS_VIEW,
    PERM_POOLS_VIEW,
    PERM_LIVE_VIEW,
    'sessions.view',
    PERM_REPORTS_VIEW,
    PERM_AUDIT_VIEW,
    PERM_RENEW,
    PERM_DISCONNECT,
    PERM_TOPUP,
    PERM_EXPORT,
  ]),
  agent: new Set([PERM_SUBSCRIBERS_VIEW, PERM_REPORTS_VIEW, PERM_RENEW]),
}

/** Base64url-decode a JWT payload segment (browser-safe, no Buffer). */
function decodeBase64Url(segment: string): string | null {
  try {
    const b64 = segment.replace(/-/g, '+').replace(/_/g, '/')
    const padded = b64 + '='.repeat((4 - (b64.length % 4)) % 4)
    return atob(padded)
  } catch {
    return null
  }
}

/**
 * Read the effective permission set from an access token's `perms` claim.
 * Returns null when the token is absent or unparseable so callers can fall back
 * to the role matrix.
 */
export function decodeTokenPerms(token: string | null | undefined): ReadonlySet<string> | null {
  if (!token) return null
  const parts = token.split('.')
  if (parts.length !== 3) return null
  const json = decodeBase64Url(parts[1])
  if (!json) return null
  try {
    const claims = JSON.parse(json) as { perms?: unknown }
    if (!Array.isArray(claims.perms)) return null
    return new Set(claims.perms.filter((p): p is string => typeof p === 'string'))
  } catch {
    return null
  }
}

/**
 * Whether a decoded permission set (or role fallback) grants a permission
 * string. The wildcard `*` (admin) grants everything.
 */
export function permsHave(perms: ReadonlySet<string>, permission: string): boolean {
  return perms.has(WILDCARD) || perms.has(permission)
}

/** Role-matrix fallback (Phase-2 tokens with no `perms` claim). */
export function hasPermission(role: string | undefined, permission: string): boolean {
  if (!role) return false
  if (role === ROLE_ADMIN) return true
  return ROLE_PERMISSIONS[role]?.has(permission) ?? false
}

File: backend/internal/auth/permissions.go
package auth

// Permission model — Phase 3 (contract C7, FR-27). Phase 2 shipped hardcoded
// role→permission sets; this phase resolves the effective set from DB-backed
// editable roles (`roles`/`role_permissions`) plus per-manager overrides
// (`manager_permission_overrides`). The frozen contract is unchanged: code
// checks *permission strings* (`<module>.<verb>` and bare action perms), never
// role names, and deny-by-default holds.
//
// Resolution happens once, at login/refresh: the effective set is embedded in
// the access token (see tokens.go) and read by Manager.Can with no per-request
// DB hit. A role or override edit therefore takes effect within one
// access-token lifetime (≤ accessTTL) — the same propagation SLA as revocation.
//
// The in-memory sets below are retained as (a) the seed source of truth the
// 0210 migration mirrors and (b) a fallback for any Manager built without a
// resolved set (e.g. unit tests, or a legacy manager row with no role_id).

import (
	"context"

	"github.com/jackc/pgx/v5/pgxpool"
)

// Built-in role names (FR-27.3). These are seeded as editable DB rows by
// migration 0210; the constants remain for the fallback path and seeding.
const (
	RoleAdmin    = "admin"
	RoleOperator = "operator"
	RoleAgent    = "agent"
)

// permWildcard is the allow-all permission held by the Admin role.
const permWildcard = "*"

// Action permissions (contract C2/C7): granted independently of module verbs.
const (
	PermRenew      = "renew"
	PermDisconnect = "disconnect"
	PermTopup      = "topup"
	PermExport     = "export"
)

// rolePermissions is the deny-by-default fallback map for non-admin builtin
// roles, mirroring the seeded role_permissions rows. Admin is allow-all and is
// intentionally absent (see roleCan). Keys are permission strings.
var rolePermissions = map[string]map[string]bool{
	RoleOperator: setOf(
		"subscribers.view", "subscribers.create", "subscribers.edit",
		"profiles.view",
		"nas.view",
		"pools.view",
		"live.view",
		"sessions.view",
		"reports.view",
		"audit.view",
		PermRenew, PermDisconnect, PermTopup, PermExport,
	),
	RoleAgent: setOf(
		"subscribers.view",
		"reports.view",
		PermRenew,
	),
}

func setOf(perms ...string) map[string]bool {
	m := make(map[string]bool, len(perms))
	for _, p := range perms {
		m[p] = true
	}
	return m
}

// roleCan reports whether a builtin role name grants a permission string. Used
// only as the fallback when a Manager carries no resolved permission set (see
// Manager.Can). Admin is unconditionally allowed; every other role is
// deny-by-default.
func roleCan(role, perm string) bool {
	if role == RoleAdmin {
		return true
	}
	return rolePermissions[role][perm]
}

// resolvePermissions computes a manager's effective permission set: the
// permissions of their assigned role, then per-manager overrides applied on top
// (granted adds, !granted removes). Returns a deduplicated, unordered slice.
//
// A manager with no role_id (legacy row) resolves from the in-memory fallback
// keyed on the legacy role text so authorization never silently opens up.
func resolvePermissions(ctx context.Context, db *pgxpool.Pool, managerID string) ([]string, error) {
	set := map[string]bool{}

	rows, err := db.Query(ctx,
		`SELECT rp.permission
		   FROM managers m
		   JOIN role_permissions rp ON rp.role_id = m.role_id
		  WHERE m.id = $1::uuid`, managerID)
	if err != nil {
		return nil, err
	}
	hadRole := false
	for rows.Next() {
		var p string
		if err := rows.Scan(&p); err != nil {
			rows.Close()
			return nil, err
		}
		set[p] = true
		hadRole = true
	}
	rows.Close()
	if err := rows.Err(); err != nil {
		return nil, err
	}

	// Legacy manager with no role_id: fall back to the builtin set by role text.
	if !hadRole {
		var role string
		if err := db.QueryRow(ctx, `SELECT role FROM managers WHERE id = $1::uuid`, managerID).Scan(&role); err == nil {
			if role == RoleAdmin {
				set[permWildcard] = true
			}
			for p := range rolePermissions[role] {
				set[p] = true
			}
		}
	}

	// Per-manager overrides layered last.
	orows, err := db.Query(ctx,
		`SELECT permission, granted FROM manager_permission_overrides WHERE manager_id = $1::uuid`, managerID)
	if err != nil {
		return nil, err
	}
	for orows.Next() {
		var p string
		var granted bool
		if err := orows.Scan(&p, &granted); err != nil {
			orows.Close()
			return nil, err
		}
		if granted {
			set[p] = true
		} else {
			delete(set, p)
		}
	}
	orows.Close()
	if err := orows.Err(); err != nil {
		return nil, err
	}

	out := make([]string, 0, len(set))
	for p := range set {
		out = append(out, p)
	}
	return out, nil
}
